File: lib/neo4j/cypher.js
module.exports = (function (R, q, request, _, bilby, discovery) {
    "use strict";

    var local = bilby.environment()
            .property("cypher", _.curry(function (query, params, transform, discovery) {
                var df = q.defer();
                request.post(discovery.data.cypher, {json: {query: query, params: params}},
                    function (err, response, body) {
                        if (err) {
                            df.reject(err);
                            return;
                        }
                        if (response.statusCode !== 200) {
                            df.reject(response);
                            return;
                        }
                        df.resolve(transform(body));
                    });
                return df.promise;
            }));

    function toJsObjs(cypherData) {
        var columns = cypherData.columns,
            rows = cypherData.data;

        return _.map(rows, function (row) { return _.zipObject(columns, row); });
    }

    function cypher(query, params, transform) {
        return discovery.services.then(local.cypher(query, params, transform));
    }

    return {
        cypherWithTransform: _.curry(cypher),
        cypher: _.curry(_.partialRight(cypher, _.identity)),
        cypherToObj: R.curryN(2, _.partialRight(cypher, toJsObjs)),
        toJsObjs: toJsObjs
    };
}(
    require("ramda"),
    require("q"),
    require("request"),
    require("lodash"),
    require("bilby"),
    require("./discovery")
));

File: fortuna/src/fortuna/cypher.clj
(ns fortuna.cypher
  (:require [clj-http.client :as client]
            [fortuna.discovery :as d]
            [fortuna.util :as util]))

(defn cypher [query, params]
  (:body (client/post (-> (d/discover) :data :cypher)
                      {:form-params {:query query, :params params}
                       :content-type :json
                       :as :json})))

(letfn [(process-cypher-datapoint [point]
          (cond (map? point) (:data point)
                (vector? point) (process-cypher-row point)
                (float? point) (util/conservative-int point)
                :else point))
        (process-cypher-row [row]
          (map process-cypher-datapoint row))]
  (defn extract-data [cypher-data-rows]
    (map process-cypher-row cypher-data-rows)))